HDDS-9824. Provide CLI that scans keys for a matching container and lists keys

mladjan-gadzic opened the pull request:
## What changes were proposed in this pull request?
HDDS-9824. Provide CLI that scans containers for keys

These changes are mostly quality of life improvements. There were use cases when it was necessary to check whether unhealthy container was empty. These changes help in that matter - this CLI outputs keys for input container ids (separated by comma). 

How is this done? Everything was done on the client side, so there is no performance impact on the server side. RocksDB tables (keyTable, fileTable and dirTable) are read. KeyTable is pretty straightforward, it is read and container id is taken from OmKeyInfo. The situation is a little different for fileTable. In order to provide key name, fileTable and dirTable have to be read, and key name needs to be constructed, because FSO buckets does not follow the same ideology as non-FSO buckets in regards to key names.

Usage example: ozone debug ldb --db=/data/metadata/om.db ckscanner -ids=1,2,3

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-9824

## How was this patch tested?
Integration tests

kerneltime commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+ */
+@CommandLine.Command(
+    name = "ckscanner",
+    description = "Parse a list of container IDs"
```
```suggestion
    description = "Find keys that reference a container"
```

kerneltime commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+      paramLabel = "containerIDs",
+      required = true,
+      description = "Set of container IDs to be used for getting all " +
+          "their keys. Example-usage: 1,11,2.(Separated by ',')")
```
Is it supposed to be `.` terminated?
```suggestion
          "their keys. Example-usage: 1,11,2 (Separated by ',')")
```

mladjan-gadzic replied:
No, that was more of a full stop, but I understand why it might be confusing. I'll remove it as suggested.

hemantk-12 replied:
I think you can end the sentence with period.
```
"their keys. Example-usage: 1,11,2 (Separated by ',').")
```

mladjan-gadzic replied:
Done.

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyInfo.java:
```diff
@@ -0,0 +1,84 @@
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
```
nit: you can use `instanceof` which will do null check as well.
```suggestion
    if (!(o instanceof ContainerKeyInfo)) {
```

mladjan-gadzic commented:
@kerneltime @hemantk-12 thanks for the review!

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+
+      // Get directory table handle
+      ColumnFamilyHandle columnFamilyHandle = getColumnFamilyHandle(
+          columnFamilyDefinition.getName().getBytes(UTF_8),
```
Do we need to get `DBDefinition` and `DBColumnFamilyDefinition`? I think you can directly pass `DIRECTORY_TABLE.getBytes(UTF_8)` rather than getting it all from `DBDefinition` and then `DBColumnFamilyDefinition`.

mladjan-gadzic replied:
You are right. Removed those.

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+        iterator.get().seekToFirst();
+        while (iterator.get().isValid()) {
+          directoryTableData.put(new String(iterator.get().key(), UTF_8),
+              ((OmDirectoryInfo) columnFamilyDefinition.getValueCodec()
```
Similar to previous comment, it can be `OmDirectoryInfo.getCodec().fromPersistedFormat(iterator.get().value())`.

mladjan-gadzic replied:
Fixed.

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+    return spec.commandLine().getOut();
+  }
+
+  public Map<Long, Path> getAbsolutePathForObjectIDs(
```
getAbsolutePathForObjectIDs is almost same as  [FSODirectoryPathResolver#getAbsolutePathForObjectIDs](https://github.com/apache/ozone/blob/15c77c6e9bb22a0bbf5d0080f81c9310051229a9/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/FSODirectoryPathResolver.java#L72). Can we just reuse existing function rather than creating new one?

mladjan-gadzic replied:
I tried that, but couldn't find a way to squeeze functionality from [FSODirectoryPathResolver#getAbsolutePathForObjectIDs](https://github.com/apache/ozone/blob/15c77c6e9bb22a0bbf5d0080f81c9310051229a9/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/FSODirectoryPathResolver.java#L72) that is needed here. There is a significant difference in accesing the tables between methods (iterators used etc.).

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+    try (ManagedRocksDB db = ManagedRocksDB.openReadOnly(dbPath,
+        columnFamilyDescriptors, columnFamilyHandles)) {
+      dbPath = removeTrailingSlashIfNeeded(dbPath);
+      DBDefinition dbDefinition = DBDefinitionFactory.getDefinition(
```
Similar to comment, I don't think it is dbDefinition is needed. You can directly access CFH.

mladjan-gadzic replied:
Removed.

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+          db.get().newIterator(columnFamilyHandle))) {
+        iterator.get().seekToFirst();
+        while (iterator.get().isValid()) {
+          directoryTableData.put(new String(iterator.get().key(), UTF_8),
```
nit: Use [StringUtils](https://github.com/apache/ozone/blob/15c77c6e9bb22a0bbf5d0080f81c9310051229a9/hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/StringUtils.java#L40) to/from byte conversion.

mladjan-gadzic replied:
Done.

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+          bucketId = Long.parseLong(keyParts[2]);
+        }
+
+        for (OmKeyLocationInfoGroup locationInfoGroup : keyLocationVersions) {
+          for (List<OmKeyLocationInfo> locationInfos :
+              locationInfoGroup.getLocationVersionMap().values()) {
+            for (OmKeyLocationInfo locationInfo : locationInfos) {
+              if (containerIds.contains(locationInfo.getContainerID())) {
+                // Generate asbolute key path for FSO keys
+                StringBuilder keyName = new StringBuilder();
+                if (tableName.equals(FILE_TABLE)) {
+                  // Load directory table only after the first fso key is found
+                  // to reduce necessary load if there are not fso keys
+                  if (!isDirTableLoaded) {
+                    long start = System.currentTimeMillis();
+                    directoryTable = getDirectoryTableData(parent.getDbPath());
+                    long end = System.currentTimeMillis();
+                    out().println(
+                        "directoryTable loaded in " + (end - start) + " ms.");
+                    isDirTableLoaded = true;
+                  }
+                  keyName.append(getFsoKeyPrefix(volumeId, bucketId, value));
+                }
+                keyName.append(value.getKeyName());
+                containerKeyInfos.add(
+                    new ContainerKeyInfo(locationInfo.getContainerID(),
+                        value.getVolumeName(), volumeId, value.getBucketName(),
+                        bucketId, keyName.toString(),
+                        value.getParentObjectID()));
+              }
+            }
+          }
+        }
```
nit: Please move this to helper function. Too much nesting.

mladjan-gadzic replied:
Fixed.

errose28 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+          "their keys. Example-usage: 1,11,2 (Separated by ',').")
+  private Set<Long> containerIds;
+
+  private static Map<String, OmDirectoryInfo> directoryTable;
```
I think the directory table may be too large to require completely loading it in to memory.

mladjan-gadzic replied:
In some use cases it might be. I wasn't sure how to handle this. Reading from table multiple times is time consuming. Any chance you have an idea how this might be handled better?

errose28 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+
+    // Get all tables from RocksDB
+    List<ColumnFamilyDescriptor> columnFamilyDescriptors =
+        RocksDBUtils.getColumnFamilyDescriptors(dbPath);
```
Can we use `OmMetadataManagerImpl` as a wrapper over the RocksDB instance? I don't think `OmMetadataManagerImpl` should actually need an `OzoneManager` instance that looks like a mistake introduced in the snapshot code.

errose28 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+                        "directoryTable loaded in " + (end - start) + " ms.");
+                    isDirTableLoaded = true;
+                  }
+                  keyName.append(getFsoKeyPrefix(volumeId, bucketId, value));
```
Looks like this is doing a full directory walk for every key found. I think this approach would be better:
1. Scan the file table and save all the matching keys.
2. If no keys in the file table match, we are done.
3. Else, do one walk through the directory table to build the full paths for all of those keys.
As it is this tool is storing all matches in memory either way. 

A more robust approach would be to have the tool print json matches as it goes so that it works with any size input and output. Then we would walk the directory and file table in one iteration to check for matches.

mladjan-gadzic replied:
There seems to be multiple approaches on how to print output. There is something like `ListVolumeHandler` for listing volumes which prints out proper JSON format and then there is something like `ListSubcommand` for listing containers which prints out JSON-like format but not true JSON. In order to have it printed out as true JSON format, data for printing needs to be kept in memory and accessed at the time of printing out.

With above in mind would you like to see proper JSON format which is better formatted and easier for end user to read or JSON-like format which is faster performance-wise but uglier for end user?

errose28 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+    List<ContainerKeyInfo> containerKeyInfos =
+        containerKeyInfoWrapper.getContainerKeyInfos();
+    if (containerKeyInfos.isEmpty()) {
+      out().println("No keys were found for container IDs: " + containerIds);
+      out().println(
```
For json tools we should only print json to stdout and print everything else to stderr otherwise it will break when piping to tools like `jq`.

mladjan-gadzic replied:
Done.

errose28 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyInfoWrapper.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Class for aggregation of collected data.
+ */
+public class ContainerKeyInfoWrapper {
```
Can we make this an inner class of `ContainerKeyScanner` since it is only used in one place to wrap a method return? Also can we combine this class with `ContaienrKeyInfoResponse` by just doing the json processing on this object?

mladjan-gadzic replied:
Done.

errose28 reviewed (commented):
Thanks for adding this debug tool @mladjan-gadzic. I left some comments inline.

mladjan-gadzic commented:
@hemantk-12  @errose28 , thank you both for the review and comments. Unfortunately, I'm currently dealing with a health crisis and won't be able to address these until the start of next week. I'll do my best to handle everything as soon as I'm able to.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+    Map<String, OmDirectoryInfo> directoryTableData = new HashMap<>();
+
+    // Get all tables from RocksDB
+    List<ColumnFamilyDescriptor> columnFamilyDescriptors =
```
I am not sure what exactly is meant by `default` here. Could you please explain?

hemantk-12 replied:
Whenever you read RoaksDB, you need to pass default CF. More details: https://github.com/facebook/rocksdb/wiki/Column-Families

 I don't think this comment is valid anymore.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/OMAdmin.java:
```diff
@@ -43,6 +40,9 @@
 
+import static org.apache.hadoop.ozone.om.OMConfigKeys.OZONE_OM_ADDRESS_KEY;
+import static org.apache.hadoop.ozone.om.OMConfigKeys.OZONE_OM_SERVICE_IDS_KEY;
+
```
Fixed.

mladjan-gadzic commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestContainerKeyScanner.java:
```diff
@@ -0,0 +1,241 @@
+  }
+
+  @Test
+  void testWhenThereAreKeysForConatainerIds() throws IOException {
```
Fixed.

mladjan-gadzic commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestContainerKeyScanner.java:
```diff
@@ -0,0 +1,241 @@
+  private PrintWriter pstdout, pstderr;
+  private CommandLine cmd;
+
+  private static final String KEYS_FOUND_OUTPUT = "{\n" +
```
Done.

mladjan-gadzic commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestContainerKeyScanner.java:
```diff
@@ -0,0 +1,241 @@
+    String volumeName = "vol1";
+    String bucketName = "bucket1";
+    // format: /volumeName/bucketName/keyName
+    String key = "/" + volumeName + "/" + bucketName + "/" + keyName;
```
Done.

mladjan-gadzic commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/OMRequestTestUtils.java:
```diff
@@ -34,6 +28,8 @@
 import org.apache.hadoop.hdds.protocol.proto.HddsProtos;
 import org.apache.hadoop.hdds.scm.pipeline.Pipeline;
 import org.apache.hadoop.hdds.scm.pipeline.PipelineID;
+import org.apache.hadoop.hdds.utils.db.cache.CacheKey;
```
Done.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+  @CommandLine.ParentCommand
+  private RDBParser parent;
+
+  @CommandLine.Option(names = {"-ids", "--container-ids"},
```
Removed.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,421 @@
+  private static CommandLine.Model.CommandSpec spec;
+
+  @CommandLine.ParentCommand
+  private RDBParser parent;
```
Done. Moved it as subcommand to `debug` command.

mladjan-gadzic commented:
@hemantk-12 @errose28 thank you for the review once more. I addressed all the review comments and left some inline questions. Please, let me know what do you think.

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,372 @@
+  public Map<Long, Path> getAbsolutePathForObjectIDs(
+      long bucketId, String prefix, Optional<Set<Long>> dirObjIds) {
+    // Root of a bucket would always have the
+    // key as /volumeId/bucketId/bucketId/
```
JFYI, this is valid only for the first dir or key without any dir.

hemantk-12 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,372 @@
+  @Override
+  public Void call() throws Exception {
+    OzoneConfiguration ozoneConfiguration = new OzoneConfiguration();
+    ozoneConfiguration.set("ozone.om.db.dirs",
```
QQ: do we need this to work for snapshot? If so, will it work with snapshot db dir?

mladjan-gadzic replied:
Initially it is not intended to work for snapshot. This can be extended in future if needed.

hemantk-12 reviewed (commented):
Thanks @mladjan-gadzic for addressing previous review comment.

I left couple of questions and some cosmetic suggestions.

adoroszlai commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestContainerKeyScanner.java:
```diff
@@ -0,0 +1,252 @@
+    closeDbStore();
+
+    int exitCode = cmd.execute(cmdArgs);
+    Assertions.assertEquals(0, exitCode);
```
nit: please `import static` assertion methods.

mladjan-gadzic replied:
Done. Furthermore, I am going to use assertj instead of junit assertions, because I do not like to mix those in a single test class.

adoroszlai commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestContainerKeyScanner.java:
```diff
@@ -0,0 +1,252 @@
+        GSON.fromJson(stdout.toString(), ContainerKeyInfoResponse.class),
+        KEYS_FOUND_OUTPUT);
+
+    Assertions.assertTrue(stderr.toString().isEmpty());
```
nit: please use `assertThat` for better failure message

```suggestion
    assertThat(stderr.toString()).isEmpty();
```

(needs `import static org.assertj.core.api.Assertions.assertThat`)

mladjan-gadzic replied:
Done.

adoroszlai commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestContainerKeyScanner.java:
```diff
@@ -0,0 +1,252 @@
+    int exitCode = cmd.execute(cmdArgs);
+    Assertions.assertEquals(0, exitCode);
+
+    Assertions.assertTrue(stderr.toString().contains(KEYS_NOT_FOUND_OUTPUT));
```
similarly:

```suggestion
    assertThat(stderr.toString()).contains(KEYS_NOT_FOUND_OUTPUT);
```

mladjan-gadzic replied:
Done.

adoroszlai commented on hadoop-ozone/tools/pom.xml:
```diff
@@ -108,6 +108,21 @@ https://maven.apache.org/xsd/maven-4.0.0.xsd">
       <artifactId>hdds-test-utils</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-params</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-junit-jupiter</artifactId>
+      <scope>test</scope>
+    </dependency>
```
These dependencies are available via `hdds-hadoop-dependency-test`.

```suggestion
```

mladjan-gadzic replied:
Done.

adoroszlai commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyInfo.java:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Class that holds basic key data in relation to container it is in.
+ */
+public class ContainerKeyInfo {
```
nit:

```suggestion
public final class ContainerKeyInfo {
```

mladjan-gadzic replied:
Done.

adoroszlai commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyInfoResponse.java:
```diff
@@ -0,0 +1,66 @@
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
```
@hemantk-12 the class must also be `final` to be OK to use `instanceof` in `equals()`.  Otherwise equality with an instance of a subclass may not be symmetric (i.e. result may vary according to which instance is `equals(other)` invoked on.

mladjan-gadzic replied:
Fixed.

adoroszlai commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,372 @@
+ * Parser for a list of container IDs, to scan for keys.
+ */
+@CommandLine.Command(
+    name = "ckscanner",
```
Please add it as a sub-command of `ozone debug container`, then it can be named something like `keys` (or `list-keys` / `find-keys`, whichever you prefer).

see `hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/container/ContainerCommands.java`

I'd avoid using the term `scan`, to avoid confusion with container scanners (data validation process in datanode).

mladjan-gadzic replied:
Done.

adoroszlai reviewed (commented):
Thanks @mladjan-gadzic for working on this.  I think this tool will be very useful.

mladjan-gadzic commented:
> Thanks @mladjan-gadzic for addressing previous review comment.
> 
> 
> 
> I left couple of questions and some cosmetic suggestions.
> 
> 
> 
> Also can you please add sample request and its output in the details? If there is any, you used for manual testing.

Thank you for the second review round. I will be sure to do that!

mladjan-gadzic commented:
> Thanks @mladjan-gadzic for working on this.  I think this tool will be very useful.

Thank you for the review! Currently I am working on something else so I am going to put this PR in draft state, but I will be sure get back to this ASAP.

mladjan-gadzic commented:
> @mladjan-gadzic thanks for working on the patch , but a quick question, why we want to implement this as a CLI feature ? We already have it in recon UI.

As far as I know, complains about Recon UI that we got on this is that Recon is a little behind the real state in cluster even after refresh. Furthermore, AFAIK there is no way to check for a specific containers/keys, but UI output shows everything, and it can be hard to look through.

devmadhuu commented:
> > @mladjan-gadzic thanks for working on the patch , but a quick question, why we want to implement this as a CLI feature ? We already have it in recon UI.
> 
> As far as I know, complains about Recon UI that we got on this is that Recon is a little behind the real state in cluster even after refresh. Furthermore, AFAIK there is no way to check for a specific containers/keys, but UI output shows everything, and it can be hard to look through.

@mladjan-gadzic , Recon works on eventual consistency concept as it has periodic tasks to run , so there might be some delay,but you can search any container id in UI itself

devmadhuu commented:
> > > > @mladjan-gadzic thanks for working on the patch , but a quick question, why we want to implement this as a CLI feature ? We already have it in recon UI.
> > > 
> > > 
> > > As far as I know, complains about Recon UI that we got on this is that Recon is a little behind the real state in cluster even after refresh. Furthermore, AFAIK there is no way to check for a specific containers/keys, but UI output shows everything, and it can be hard to look through.
> > 
> > 
> > @mladjan-gadzic , Recon works on eventual consistency concept as it has periodic tasks to run , so there might be some delay,but you can search any container id in UI itself
> 
> Could you please point me to that page/endpoint? As I can remember, issue was that there was pagination, and in order to get to a particular key we'd need to know on which page it is excatly if there is large number of keys.

**Recon UI currently search across all pages in pagination is under development. But you can still use api/v1/containers/<container_id>/keys**  using curl command

errose28 commented:
I'm +1 for adding a CLI even if Recon has a similar feature. We've also seen instances where there are issues with Recon so it would be good to have a way to check OM RocksDB directly.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,372 @@
+    return Pair.of(Long.parseLong(keyParts[1]), Long.parseLong(keyParts[2]));
+  }
+
+  private void processKeyData(
```
If I am not mistaken, non FSO layouts does not have volume/bucket id (it is 0), so I'd keep it here as well.

hemantk-12 replied:
ObjectId (volumeId and BucketId) still should be there no matter their type OBS or FSO.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,372 @@
+    return path;
+  }
+
+  private String getFsoKeyPrefix(long volumeId, long bucketId,
```
Fixed.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,372 @@
+
+  private String getFsoKeyPrefix(long volumeId, long bucketId,
+                                 OmKeyInfo value) {
+    String prefix =
```
Done.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,372 @@
+
+    printOutput(containerKeyInfoResponse);
+
+    closeStdChannels();
```
Fixed.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyInfoResponse.java:
```diff
@@ -0,0 +1,66 @@
+public class ContainerKeyInfoResponse {
+
+  private final long keysProcessed;
+  private final Map<Long, List<ContainerKeyInfo>> containerKeys;
```
Fixed.

mladjan-gadzic commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestContainerKeyScanner.java:
```diff
@@ -0,0 +1,241 @@
+ * This class tests `ozone debug ldb ckscanner` CLI that reads from RocksDB
+ * and gets keys for container ids.
+ */
+public class TestContainerKeyScanner {
```
I've managed to incorporate all of the above conditions into three tests.

mladjan-gadzic commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/ContainerKeyScanner.java:
```diff
@@ -0,0 +1,372 @@
+    }
+  }
+
+  private void processFileData(
```
> We recently increased chars per line to 120. Please update your IDE setting if you have not.

Thanks for this! Somehow I've missed this update.

> Use early return to avoid nesting.

Fixed.

> I think getDirectoryTableData can be loaded outside the loops and no need to use isDirTableLoaded to check if it is loaded or not.

You are right, however, I used that to enable quick lazy loading mechanism.

> If it is to load only when needed, may be use some other way to load lazily.

Do you have any suggestions for this?

hemantk-12 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestFindContainerKeys.java:
```diff
@@ -0,0 +1,286 @@
+    createFile(volumeId, bucketId, "key4", -984L, dirObjectId2, 3L);
+    createFile(volumeId, bucketId, "key5", -983L, dirObjectId2, 4L);
+
+    closeDbStore();
```
Why do we need to close it in individual test when we are closing it in `@AfterEach`?

hemantk-12 reviewed (commented):
Overall looks good to me.

Thanks for the patch @mladjan-gadzic.

aswinshakil commented:
@mladjan-gadzic There are some merge conflicts, Can you rebase this PR with the latest master? We can work on getting this merged. Thank you for your contribution.

mladjan-gadzic commented:
> @mladjan-gadzic There are some merge conflicts, Can you rebase this PR with the latest master? We can work on getting this merged. Thank you for your contribution.

hi @aswinshakil! i am working on something else right now, and to be honest i've lost context on this so i'd need some time to get back to this. if time allows, i'd get back to it in the future, but i cannot promise anything.

errose28 commented:
Closing this for now in favor of HDDS-11747 since @sumitagrawl has started working on another implementation.
